Concept Doc: Anatomy of an SDK app (#4426)
cosmos/cosmos-sdk: docs/concepts/app-anatomy.md

# Anatomy of an SDK Application

## Pre-requisite reading

- [High-level overview of the architecture of an SDK application](../intro/sdk-app-architecture.md)
- [Cosmos SDK design overview](../intro/sdk-design.md)

## Synopsis

This document describes the core parts of a Cosmos SDK application. The placeholder name for this application will be `app`.

- [Node Client](#node-client)
- [Core Application File](#core-application-file)
- [Modules](#modules)

> - [Intefaces](#interfaces)

Reviewer: ```suggestion
- [Interfaces](#interfaces)
```

├── go.mod
└── Makefile
``` 

## Node Client (Daemon)

The Daemon, or Full-Node Client, is the core process of an SDK-based blockchain. Participants in the network run this process to initialize their state-machine, connect with other full-nodes and update their state-machine as new blocks come in. 

```
                ^  +-------------------------------+  ^
                |  |                               |  |
                |  |  State+machine = Application  |  |
                |  |                               |  |   Built with Cosmos SDK
                |  |            ^      +           |  |
                |  +----------- | ABCI | ----------+  v
                |  |            +      v           |  ^
                |  |                               |  |
Blockchain Node |  |           Consensus           |  |
                |  |                               |  |
                |  +-------------------------------+  |   Tendermint Core
                |  |                               |  |
                |  |           Networking          |  |
                |  |                               |  |
                v  +-------------------------------+  v
```

> The blockchain full-node presents itself as a binary, generally suffixed by `-d` (e.g. `appd` for `app` or `gaiad` for the `gaia`) for "daemon". This binary is built by running a simple `main.go` function placed in `cmd/appd/`. This operation usually happens through the [Makefil](#dependencies-and-makefile).

Reviewer: ```suggestion
The blockchain full-node presents itself as a binary, generally suffixed by `-d` (e.g. `appd` for `app` or `gaiad` for the `gaia`) for "daemon". This binary is built by running a simple `main.go` function placed in `cmd/appd/`. This operation usually happens through the [Makefile](#dependencies-and-makefile).
```

1- Create an instance of the state-machine defined in [`app.go`](#core-application-file) using the `appCreator`. 
2- Initialize the state-machine with the latest known state, extracted from the `db` stored in the `~/.appd/data` folder. At this point, the state-machine is at height `appBlockHeight`. 
3- Create and start a new Tendermint instance. Among other things, the node will perform a handshake with its peers. It will get the latest `blockHeight` from them, and replay blocks to sync to this height if it is greater than the local `appBlockHeight`. If `appBlockHeight` is `0`, the node is starting from genesis and Tendermint sends an `InitChain` message via the ABCI to the `app`, which triggers the [`InitChainer`](#initchainer).

To learn more about the `start` command, [click here](./node.md#start-command).

## Core Application File

In general, the core of the state-machine is defined in a file called `app.go`. It mainly contains the **type definition of the application** and functions to **create and initialize it**. 

### Type Definition of the Application

The first thing defined in `app.go` is the `type` of the application. It is generally comprised of the following parts:

> - **A reference to [`baseapp`](./baseapp.md).** The custom application defined in `app.go` is a golang embedding of the `baseapp` type. `baseapp` implements most of the core logic for the application, including all the [ABCI methods](https://tendermint.com/docs/spec/abci/abci.html#overview) and the routing logic. When a transaction is relayed by Tendermint to the application, the latter uses `baseapp`'s methods to route them to the appropriate module. 

Reviewer: ```suggestion
- **A reference to [`baseapp`](./baseapp.md).** The custom application defined in `app.go` is a Golang embedding of the `baseapp` type. `baseapp` implements most of the core logic for the application, including all the [ABCI methods](https://tendermint.com/docs/spec/abci/abci.html#overview) and the routing logic. When a transaction is relayed by Tendermint to the application, the latter uses `baseapp`'s methods to route them to the appropriate module. 
```

- **A list of module's `keepers`.** Each module defines an abstraction called `keeper`, which handles reads and writes for this module's store(s). The `keeper`'s methods of one module can be called from other modules (if authorized), which is why they are declared in the application's type. 
- **A reference to a `codec`.** The Cosmos SDK gives developers the freedom to choose the encoding framework for their application. The application's `codec` is used to serialize and deserialize data structures in order to store them, as stores can only persist `[]bytes`. The `codec` must be deterministic. Most SDK application use [amino](./amino.md) as their `codec`. 

You can see an example of application type definition [here](https://github.com/cosmos/sdk-application-tutorial/blob/master/app.go#L27-L43).

### Constructor Function

> This function constructs a new application of the type defined above. It is [called](https://github.com/cosmos/cosmos-sdk/blob/master/server/start.go#L117) everytime the full-node is started with the `start` command. Here are the main actions performed by this function:

Reviewer: ```suggestion
This function constructs a new application of the type defined above. It is [called](https://github.com/cosmos/cosmos-sdk/blob/master/server/start.go#L117) every time the full-node is started with the `start` command. Here are the main actions performed by this function:
```

> - Instanciate a new application with a reference to a `baseapp` instance, a codec and all the appropriate store keys.

Reviewer: ```suggestion
- Instantiate a new application with a reference to a `baseapp` instance, a codec and all the appropriate store keys.
```

> - Instanciate all the `keepers` defined in the application's `type`.

Reviewer: ```suggestion
- Instantiate all the `keepers` defined in the application's `type`.
```

- Set the application's [`initChainer`](#initchainer) and mount the stores. 
- Return the application. 

Note that this function only creates an instance of the app, while the actual state is either carried over from the `~/.appd/data` folder if the node is restarted, or generated from the genesis file if the node is started for the first time. 

You can see an example of application constructor [here](https://github.com/cosmos/sdk-application-tutorial/blob/master/app.go#L46-L128).

### InitChainer

The `initChainer` is a function that initializes the state of the application from a [genesis file](./genesis.md) (i.e. token balances of genesis accounts). It is called when the application received the `InitChain` message from the Tendermint engine, which happens when the node is started at `appBlockHeight == 0` (i.e. on genesis). The application must set the `initChainer` in its constructor via the [`setInitChainer`](https://godoc.org/github.com/cosmos/cosmos-sdk/baseapp#BaseApp.SetInitChainer) method. 

You can see an example of an `initChainer` [here](https://github.com/cosmos/sdk-application-tutorial/blob/master/app.go#L137-L155).

### Register Codec

> The `MakeCodec` function is the last important function of the `app.go` file. The goal of this function is to instanciate a codec `cdc` (e.g. [amino](./amino.md)) and calls the `RegisterCodec(*codec.Codec)` method of each module used within the application to register `cdc` to each module. 

Reviewer: ```suggestion
The `MakeCodec` function is the last important function of the `app.go` file. The goal of this function is to instantiate a codec `cdc` (e.g. [amino](./amino.md)) and calls the `RegisterCodec(*codec.Codec)` method of each module used within the application to register `cdc` to each module. 
```

You can see an example of a `MakeCodec` [here](You can see an example of an `initChainer` [here](https://github.com/cosmos/sdk-application-tutorial/blob/master/app.go#L189-L198).).

## Modules

Modules are the heart and soul of an SDK application. They can be considered as state-machines within the state-machine. When a transaction is relayed from the underlying Tendermint engine via the ABCI to the application, it is routed by `baseapp` to the appropriate module in order to be processed. This paradigm enables developers to easily build complex state-machines, as most of the modules they need often already exist. For developers, most of the work involved in building an SDK application revolves around building custom modules required by their application that do not exist, and integrating them with modules that do already exist into one coherent application. In the application directory, the standard practice is to store modules in the `x/` folder. 

To learn more about modules, [click here](./modules.md)

### Message Types

> A message is a custom type defined by each module that implements the [`message`](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx_msg.go#L8-L29) interface. Each `transaction` contains one or multiple `messages`. When a valid block of transactions is received by the full-node, Tendermint relays each one to the application via [`DeliverTx`](https://tendermint.com/docs/app-dev/abci-spec.html#delivertx). Upon receiving the transaction, the application first unmarshalls it. Then, it extracts the message(s) contained in the application. With the [`Type()`](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx_msg.go#L16) method, `baseapp` is able to know which modules defines the message. It is then able to route it to the appropriate module's [handler](#handler) in order for the message to be processed. If the message is succesfully processed, the state is updated. 

Reviewer: ```suggestion
A message is a custom type defined by each module that implements the [`message`](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx_msg.go#L8-L29) interface. Each `transaction` contains one or multiple `messages`. When a valid block of transactions is received by the full-node, Tendermint relays each one to the application via [`DeliverTx`](https://tendermint.com/docs/app-dev/abci-spec.html#delivertx). Upon receiving the transaction, the application first unmarshalls it. Then, it extracts the message(s) contained in the application. With the [`Type()`](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx_msg.go#L16) method, `baseapp` is able to know which modules defines the message. It is then able to route it to the appropriate module's [handler](#handler) in order for the message to be processed. If the message is successfully processed, the state is updated. 
```

Module developers create custom message types when they build their own module. The general practice is to prefix the type declaration of the message with `Msg`. For example, the message type `MsgSend` allows users to transfer tokens. It is processed by the handler of the `bank` module, which ultimately calls the `keeper` of the `auth` module in order to update the state.

To learn more about messages, [click here](./tx-msgs.md)

### Handler

> The `handler` refers to the part of the module responsible for processing the message after it is routed by `baseapp`. `handler` functions of modules (except those of the `auth` module) are only executed if the transaction is relayed from Tendermint by the `DeliverTx` ABCI message. If the transaction is realyed by `CheckTx`, only stateless checks and fee-related (i.e. `auth` module-related) stateful checks are performed. To better understand the difference between `DeliverTx`and `CheckTx`, as well as the difference between stateful and stateless checks, click [here](./tx-lifecycle.md).

Reviewer: ```suggestion
The `handler` refers to the part of the module responsible for processing the message after it is routed by `baseapp`. `handler` functions of modules (except those of the `auth` module) are only executed if the transaction is relayed from Tendermint by the `DeliverTx` ABCI message. If the transaction is relayed by `CheckTx`, only stateless checks and fee-related (i.e. `auth` module-related) stateful checks are performed. To better understand the difference between `DeliverTx`and `CheckTx`, as well as the difference between stateful and stateless checks, click [here](./tx-lifecycle.md).
```

The handler of a module is generally defined in a file called `handler.go` and consists of:

- A **switch function** `NewHandler` to route the message to the appropriate handler function. This function returns a `handler` function, and is used in `app.go` to initialize the [application's router](./baseapp.md#routing). See an example of such a switch [here](https://github.com/cosmos/sdk-application-tutorial/blob/master/x/nameservice/handler.go#L10-L22).
- **One handler function for each message type defined by the module**. Developers write the message processing logic in these functions. This generally involves doing stateful checks to ensure the message is valid and calling [`keeper`](#keeper)'s methods to update the state. 

> Handler functions return a result of type [`sdk.Result`](https://github.com/cosmos/cosmos-sdk/blob/master/types/result.go#L14-L37), which informs the application on wether the message was succesfully processed and.

Reviewer: ```suggestion
Handler functions return a result of type [`sdk.Result`](https://github.com/cosmos/cosmos-sdk/blob/master/types/result.go#L14-L37), which informs the application on wether the message was successfully processed and.
```

To learn more about handlers, [click here](./handler.md).

### Keeper

`Keepers` are the gatekeepers of their module's store(s). To read or write in a module's store, it is mandatory to go through one of its `keeper`'s methods. This is ensured by the [object-capabilities](./ocap.md) model of the Cosmos SDK. Only objects that hold the key to a store can access it, and only the module's `keeper` should hold the key(s) to the module's store(s).

`Keepers` are generally defined in a file called `keeper.go`. It contains the `keeper`'s type definition and methods. 

The `keeper` type definition generally consists of:

- **Key(s)** to the module's store(s) in the multistore. 
- Reference to **other module's `keepers`**. Only needed if the `keeper` needs to access other module's store(s) (either to read or write from them).

> - A reference to the application's **codec**. The `keeper` needs it to marshal structs before storing them, or to unmarhsal them when it retrieves them, because stores only accept `[]bytes` as value. 

Reviewer: ```suggestion
- A reference to the application's **codec**. The `keeper` needs it to marshal structs before storing them, or to unmarshal them when it retrieves them, because stores only accept `[]bytes` as value. 
```

- A `RegisterRoutes` function, which registers each route defined in the file. This function is called from the [main application's interface](#application-interfaces) for each module used within the application. The router used in the SDK is [Gorilla's mux](https://github.com/gorilla/mux).
- Custom request type definitions for each query or transaction creation function that needs to be exposed. These custom request types build on the [base `request` type](https://github.com/cosmos/cosmos-sdk/blob/master/types/rest/rest.go#L32-L43) of the Cosmos SDK. 
- One handler function for each request that can be routed to the given module. These functions implement the core logic necessary to serve the request.

See an example of a module's `rest.go` file [here](https://github.com/cosmos/sdk-application-tutorial/blob/master/x/nameservice/client/rest/rest.go).

To learn more about modules REST interface, [click here](./module-interfaces.md#rest).

## Application Interface

> Interfaces let end-users interract with full-node clients. This means querying data from the full-node or creating and sending new transactions to be relayed by the full-node and eventually included in a block. 

Reviewer: ```suggestion
Interfaces let end-users interact with full-node clients. This means querying data from the full-node or creating and sending new transactions to be relayed by the full-node and eventually included in a block. 
```

# Anatomy of an SDK Application

## Pre-requisite reading

- [High-level overview of the architecture of an SDK application](../intro/sdk-app-architecture.md)
- [Cosmos SDK design overview](../intro/sdk-design.md)

## Synopsis

This document describes the core parts of a Cosmos SDK application. The placeholder name for this application will be `app`.

- [Node Client](#node-client)
- [Core Application File](#core-application-file)
- [Modules](#modules)
- [Interfaces](#interfaces)
- [Dependencies and Makefile](#dependencies-and-makefile)

The core parts listed above will generally translate to the following directory tree:

```
./app
├── cmd/
│   ├── appd
│   └── appcli
├── app.go
├── x/

> │   ├── auth

Reviewer 2: ```suggestion
│   ├── auth
│   ├── ...
```

│   └── bank

> ├── go.mod

Reviewer 2: ```suggestion
├── go.mod
├── go.sum
```

Author: Agree with @rigelrozanski, I think `go.mod` is sufficient. Maybe it's implicit, but does not hurt to have it nonetheless

Reviewer 2: If we're stating what a standard application _should_ look like as far as structure, then I do think it should be included.

                v  +-------------------------------+  v
```
The blockchain full-node presents itself as a binary, generally suffixed by `-d` (e.g. `appd` for `app` or `gaiad` for the `gaia`) for "daemon". This binary is built by running a simple `main.go` function placed in `cmd/appd/`. This operation usually happens through the [Makefile](#dependencies-and-makefile).

To learn more about the `main.go` function, [click here](./node.md#main-function).

Once the main binary is built, the node can be started by running the `start` command. The core logic behind the `start` command is implemented in the SDK itself in the [`/server/start.go`](https://github.com/cosmos/cosmos-sdk/blob/master/server/start.go) file. The main [`start` command function](https://github.com/cosmos/cosmos-sdk/blob/master/server/start.go#L31) takes a [`context`](https://godoc.org/github.com/cosmos/cosmos-sdk/client/context) and [`appCreator`](#constructor-function-(`appCreator`)) as arguments. The `appCreator` is a constructor function for the SDK application, and is used in the starting process of the full-node. 

The `start` command function primarily does three things:

> 1- Create an instance of the state-machine defined in [`app.go`](#core-application-file) using the `appCreator`. 

Reviewer 2: ```suggestion
1. Create an instance of the state-machine defined in [`app.go`](#core-application-file) using the `appCreator`. 
```

> 2- Initialize the state-machine with the latest known state, extracted from the `db` stored in the `~/.appd/data` folder. At this point, the state-machine is at height `appBlockHeight`. 

Reviewer 2: ```suggestion
2. Initialize the state-machine with the latest known state, extracted from the `db` stored in the `~/.appd/data` folder. At this point, the state-machine is at height `appBlockHeight`. 
```

> 3- Create and start a new Tendermint instance. Among other things, the node will perform a handshake with its peers. It will get the latest `blockHeight` from them, and replay blocks to sync to this height if it is greater than the local `appBlockHeight`. If `appBlockHeight` is `0`, the node is starting from genesis and Tendermint sends an `InitChain` message via the ABCI to the `app`, which triggers the [`InitChainer`](#initchainer).

Reviewer 2: ```suggestion
3. Create and start a new Tendermint instance. Among other things, the node will perform a handshake with its peers. It will get the latest `blockHeight` from them, and replay blocks to sync to this height if it is greater than the local `appBlockHeight`. If `appBlockHeight` is `0`, the node is starting from genesis and Tendermint sends an `InitChain` message via the ABCI to the `app`, which triggers the [`InitChainer`](#initchainer).
```

To learn more about the `start` command, [click here](./node.md#start-command).

## Core Application File

In general, the core of the state-machine is defined in a file called `app.go`. It mainly contains the **type definition of the application** and functions to **create and initialize it**. 

### Type Definition of the Application

The first thing defined in `app.go` is the `type` of the application. It is generally comprised of the following parts:

> - **A reference to [`baseapp`](./baseapp.md).** The custom application defined in `app.go` is a Golang embedding of the `baseapp` type. `baseapp` implements most of the core logic for the application, including all the [ABCI methods](https://tendermint.com/docs/spec/abci/abci.html#overview) and the routing logic. When a transaction is relayed by Tendermint to the application, the latter uses `baseapp`'s methods to route them to the appropriate module. 

Reviewer 3: it might be nice to use the language "the app extends baseapp" - because that's a technically correct way to describe what the app struct is doing.

Author: ok

> - **A list of store keys**. The [store](./store.md), which contains the entire state, is implemented as a multistore (i.e. a store of stores) in the Cosmos SDK. Each module uses one or multiple stores in the multistore to persist their part of the state. These stores can be accessed with specific keys that are declared in the `app` type. These keys, along with the `keepers`, are at the heart of the [object-capabilities model](../intro/ocap.md) of the Cosmos SDK.  

Author: not sure, I think we've always called it a multistore so...

> - **A list of store keys**. The [store](./store.md), which contains the entire state, is implemented as a multistore (i.e. a store of stores) in the Cosmos SDK. Each module uses one or multiple stores in the multistore to persist their part of the state. These stores can be accessed with specific keys that are declared in the `app` type. These keys, along with the `keepers`, are at the heart of the [object-capabilities model](../intro/ocap.md) of the Cosmos SDK.  

Reviewer 4: 🤷‍♂ @rigelrozanski thoughts ?

> - **A list of store keys**. The [store](./store.md), which contains the entire state, is implemented as a multistore (i.e. a store of stores) in the Cosmos SDK. Each module uses one or multiple stores in the multistore to persist their part of the state. These stores can be accessed with specific keys that are declared in the `app` type. These keys, along with the `keepers`, are at the heart of the [object-capabilities model](../intro/ocap.md) of the Cosmos SDK.  

Reviewer 2: It is a multi-store -- a store of stores. The baseapp contains the core multi-store and in that, is a list of sub-stores (eg. auth, staking, etc...)

- **A reference to a `codec`.** The Cosmos SDK gives developers the freedom to choose the encoding framework for their application. The application's `codec` is used to serialize and deserialize data structures in order to store them, as stores can only persist `[]bytes`. The `codec` must be deterministic. Most SDK application use [amino](./amino.md) as their `codec`. 

You can see an example of application type definition [here](https://github.com/cosmos/sdk-application-tutorial/blob/master/app.go#L27-L43).

### Constructor Function

This function constructs a new application of the type defined above. It is [called](https://github.com/cosmos/cosmos-sdk/blob/master/server/start.go#L117) every time the full-node is started with the `start` command. Here are the main actions performed by this function:

- Instanciate a new application with a reference to a `baseapp` instance, a codec and all the appropriate store keys.

> - Instantiate all the `keepers` defined in the application's `type`.

Reviewer 3: ```suggestion
- Instantiate all module `keeper`s defined in the application's `type`.
```

> - Instantiate all the `keepers` defined in the application's `type`.

Reviewer 3: might be good to add a point (or subpoint) regarding all the specialized module functionality which is defined with a new app, such as setting invariants (crisis) or adding custom proposal handlers client functionality (gov) - all the things which tie the keepers together

Reviewer 4: ++

Author: Yes, good point. Actually I am going to rework the doc a bit to include the new logic of the module manager

> - Initialize the application's [`routes`](./baseapp.md#routing) with the [`handlers`](#handler) of each one of the application's modules. When a transaction is relayed to the application by Tendermint via the ABCI, it is routed to the appropriate module's handler using the routes defined here. 

Reviewer 4: ```suggestion
- Initialize the application's [`transaction routes`](./baseapp.md#routing) with the [`handlers`](#handler) of each one of the application's modules. When a transaction is relayed to the application by Tendermint via the ABCI, it is routed to the appropriate module's handler using the routes defined here. 
```

- Initialize the application's [query routes](./baseapp.md#query-routing) with the [`queriers`](#querier) of each of the application's modules. When a user query comes in, it is routed to the appropriate module using the query routes defined here. 
- Set the application's [`initChainer`](#initchainer) and mount the stores. 
- Return the application. 

Note that this function only creates an instance of the app, while the actual state is either carried over from the `~/.appd/data` folder if the node is restarted, or generated from the genesis file if the node is started for the first time. 

You can see an example of application constructor [here](https://github.com/cosmos/sdk-application-tutorial/blob/master/app.go#L46-L128).

### InitChainer

> The `initChainer` is a function that initializes the state of the application from a [genesis file](./genesis.md) (i.e. token balances of genesis accounts). It is called when the application received the `InitChain` message from the Tendermint engine, which happens when the node is started at `appBlockHeight == 0` (i.e. on genesis). The application must set the `initChainer` in its constructor via the [`setInitChainer`](https://godoc.org/github.com/cosmos/cosmos-sdk/baseapp#BaseApp.SetInitChainer) method. 

Reviewer 4: ```suggestion
The `initChainer` is a function that initializes the state of the application from a [genesis file](./genesis.md) (i.e. genesis accounts and their balance,  modules' genesis state, etc). It is called when the application received the `InitChain` message from the Tendermint engine, which happens when the node is started at `appBlockHeight == 0` (i.e. on genesis). The application must set the `initChainer` in its constructor via the [`setInitChainer`](https://godoc.org/github.com/cosmos/cosmos-sdk/baseapp#BaseApp.SetInitChainer) method. 
```

You can see an example of an `initChainer` [here](https://github.com/cosmos/sdk-application-tutorial/blob/master/app.go#L137-L155).

### Register Codec

The `MakeCodec` function is the last important function of the `app.go` file. The goal of this function is to instantiate a codec `cdc` (e.g. [amino](./amino.md)) and calls the `RegisterCodec(*codec.Codec)` method of each module used within the application to register `cdc` to each module. 

> In turn, the `RegisterCodec` function of each module register the custom interfaces and type structures of their respective module so that they can be marhsaled and unmarshaled. 

Reviewer 4: ```suggestion
In turn, the `RegisterCodec` function of each module register the custom interfaces and type structures of their respective module so that they can be marshaled and unmarshaled. 
```

> You can see an example of a `MakeCodec` [here](You can see an example of an `initChainer` [here](https://github.com/cosmos/sdk-application-tutorial/blob/master/app.go#L189-L198).).

Reviewer 4: ```suggestion
You can see an example of a `MakeCodec` [here](https://github.com/cosmos/sdk-application-tutorial/blob/master/app.go#L189-L198).
```

## Modules

> Modules are the heart and soul of an SDK application. They can be considered as state-machines within the state-machine. When a transaction is relayed from the underlying Tendermint engine via the ABCI to the application, it is routed by `baseapp` to the appropriate module in order to be processed. This paradigm enables developers to easily build complex state-machines, as most of the modules they need often already exist. For developers, most of the work involved in building an SDK application revolves around building custom modules required by their application that do not exist, and integrating them with modules that do already exist into one coherent application. In the application directory, the standard practice is to store modules in the `x/` folder. 

Reviewer 2: ```suggestion
Modules are the heart and soul of an SDK application. They can be considered as state-machines within the state-machine. When a transaction is relayed from the underlying Tendermint engine via ABCI to the application, it is routed by `baseapp` to the appropriate module in order to be processed. This paradigm enables developers to easily build complex state-machines, as most of the modules they need often already exist. For developers, most of the work involved in building an SDK application revolves around building custom modules required by their application that do not exist, and integrating them with modules that do already exist into one coherent application. In the application directory, the standard practice is to store modules in the `x/` folder. 
```

Reviewer 3: So sdk modules are stored in the _the_ `x/` folder but for a custom applications they would be stored in _a_ `x/` folder on the root level of your application... might be good to make this distinction here

To learn more about modules, [click here](./modules.md)

### Message Types

> A message is a custom type defined by each module that implements the [`message`](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx_msg.go#L8-L29) interface. Each `transaction` contains one or multiple `messages`. When a valid block of transactions is received by the full-node, Tendermint relays each one to the application via [`DeliverTx`](https://tendermint.com/docs/app-dev/abci-spec.html#delivertx). Upon receiving the transaction, the application first unmarshalls it. Then, it extracts the message(s) contained in the application. With the [`Type()`](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx_msg.go#L16) method, `baseapp` is able to know which modules defines the message. It is then able to route it to the appropriate module's [handler](#handler) in order for the message to be processed. If the message is successfully processed, the state is updated. 

Reviewer 4: ```suggestion
A message is a custom type defined by each module that implements the [`Msg`](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx_msg.go#L8-L29) interface. Each `transaction` contains one or multiple `messages`. When a valid block of transactions is received by the full-node, Tendermint relays each one to the application via [`DeliverTx`](https://tendermint.com/docs/app-dev/abci-spec.html#delivertx). Upon receiving the transaction, the application first unmarshalls it. Then, it extracts the message(s) contained in the application. With the [`Type()`](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx_msg.go#L16) method, `baseapp` is able to know which modules defines the message. It is then able to route it to the appropriate module's [handler](#handler) in order for the message to be processed. If the message is successfully processed, the state is updated. 
```

Reviewer 4: > When a valid block of transactions is received by the full-node, Tendermint relays each one to the application via [`DeliverTx`](https://tendermint.com/docs/app-dev/abci-spec.html#delivertx). Upon receiving the transaction, the application first unmarshalls it. Then, it extracts the message(s) contained in the application. With the [`Type()`](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx_msg.go#L16) method, `baseapp` is able to know which modules defines the message. It is then able to route it to the appropriate module's [handler](#handler) in order for the message to be processed. If the message is successfully processed, the state is updated.

this should be a list of steps instead

> Module developers create custom message types when they build their own module. The general practice is to prefix the type declaration of the message with `Msg`. For example, the message type `MsgSend` allows users to transfer tokens. It is processed by the handler of the `bank` module, which ultimately calls the `keeper` of the `auth` module in order to update the state.

Reviewer 4: ```suggestion
Module developers create custom message types when they build their own module. The general practice is to prefix the type declaration of the message with `Msg`. For example, the message type `MsgSend` allows users to transfer tokens. It is processed by the handler of the `bank` module, which ultimately calls the `keeper` of the `auth` module in order to update the accounts' state.
```

To learn more about messages, [click here](./tx-msgs.md)

### Handler

The `handler` refers to the part of the module responsible for processing the message after it is routed by `baseapp`. `handler` functions of modules (except those of the `auth` module) are only executed if the transaction is relayed from Tendermint by the `DeliverTx` ABCI message. If the transaction is relayed by `CheckTx`, only stateless checks and fee-related (i.e. `auth` module-related) stateful checks are performed. To better understand the difference between `DeliverTx`and `CheckTx`, as well as the difference between stateful and stateless checks, click [here](./tx-lifecycle.md).

The handler of a module is generally defined in a file called `handler.go` and consists of:

> - A **switch function** `NewHandler` to route the message to the appropriate handler function. This function returns a `handler` function, and is used in `app.go` to initialize the [application's router](./baseapp.md#routing). See an example of such a switch [here](https://github.com/cosmos/sdk-application-tutorial/blob/master/x/nameservice/handler.go#L10-L22).

Reviewer 4: this example probably shouldn't be the tutorial handler as it is continuously being updated, maybe something more stable as `staking` or `bank`?

Author: Well it's updated according to the latest best practices, so I think it's a valid example.

Reviewer 4: fair enough

- **One handler function for each message type defined by the module**. Developers write the message processing logic in these functions. This generally involves doing stateful checks to ensure the message is valid and calling [`keeper`](#keeper)'s methods to update the state. 

> Handler functions return a result of type [`sdk.Result`](https://github.com/cosmos/cosmos-sdk/blob/master/types/result.go#L14-L37), which informs the application on wether the message was successfully processed and.

Reviewer 4: ```suggestion
Handler functions return a result of type [`sdk.Result`](https://github.com/cosmos/cosmos-sdk/blob/master/types/result.go#L14-L37), which informs the application on whether the message was successfully processed.
```

> wether: A castrated male sheep or goat. 😄

To learn more about `keepers`, [click here](./keeper.md).

### Querier 

`Queriers` are very similar to `handlers`, except they serve user queries to the state as opposed to processing transactions. A query is initiated from an [interface](#intefaces) by an end-user who provides a `queryRoute` and some `data`. The query is then routed to the correct application's `querier` by `baseapp`'s [`handleQueryCustom`](https://github.com/cosmos/cosmos-sdk/blob/master/baseapp/baseapp.go#L519-L556) method using `queryRoute`. 

The `Querier` of a module are defined in a file called `querier.go`, and consists of:

- A **switch function** `NewQuerier` to route the query to the appropriate `querier` function. This function returns a `querier` function, and is used in `app.go` to initialize the [application's query router](./baseapp.md#query-routing). See an example of such a switch [here](https://github.com/cosmos/sdk-application-tutorial/blob/master/x/nameservice/querier.go#L21-L34).

> - - **One querier function for each data type defined by the module that needs to be queryable**. Developers write the query processing logic in these functions. This generally involves calling [`keeper`](#keeper)'s methods to query the state and marshalling it to JSON. See an example of `querier` functions [here](https://github.com/cosmos/sdk-application-tutorial/blob/master/x/nameservice/querier.go#L37-L101).

Reviewer 2: ```suggestion
- **One querier function for each data type defined by the module that needs to be queryable**. Developers write the query processing logic in these functions. This generally involves calling [`keeper`](#keeper)'s methods to query the state and marshalling it to JSON. See an example of `querier` functions [here](https://github.com/cosmos/sdk-application-tutorial/blob/master/x/nameservice/querier.go#L37-L101).
```

- **A list of store keys**. The [store](./store.md), which contains the entire state, is implemented as a multistore (i.e. a store of stores) in the Cosmos SDK. Each module uses one or multiple stores in the multistore to persist their part of the state. These stores can be accessed with specific keys that are declared in the `app` type. These keys, along with the `keepers`, are at the heart of the [object-capabilities model](../intro/ocap.md) of the Cosmos SDK.  
- **A list of module's `keepers`.** Each module defines an abstraction called `keeper`, which handles reads and writes for this module's store(s). The `keeper`'s methods of one module can be called from other modules (if authorized), which is why they are declared in the application's type and exported as interfaces to other modules so that they are only allowed to access the authorized functions. 

> - **A reference to a `codec`.** The Cosmos SDK gives developers the freedom to choose the encoding framework for their application. The application's `codec` is used to serialize and deserialize data structures in order to store them, as stores can only persist `[]bytes`. The `codec` must be deterministic. Most SDK application use [amino](./amino.md) as their `codec`. 

Reviewer 5: I would rephrase `Most SDK application use amino as their codec` into `The default codec is amino`

Reviewer 4: ++

- Instantiate a new application with a reference to a `baseapp` instance, a codec and all the appropriate store keys.
- Instantiate all the [`keeper`s](#keeper) defined in the application's `type` using the `NewKeeper` function of each of the application's modules. Note that `keepers` must be instantiated in the correct order, as the `NewKeeper` of one module might require a reference to another module's `keeper`. 
- Instantiate the application's [module manager](./module-manager.md) with the [`AppModule`](#application-module-interface) object of each of the application's modules. 
- With the module manager, initialize the application's [`routes`](./baseapp.md#routing) and [query routes](./baseapp.md#query-routing). When a transaction is relayed to the application by Tendermint via the ABCI, it is routed to the appropriate module's [`handler`](#handler) using the routes defined here. Likewise, when a query is received by the application, it is routed to the appropriate module's [`querier`](#querier) using the query routes defined here. 

> - With the module manager, register the [application's modules' invariants](./invariants.md). Invariants are variables (e.g. total supply of a token) that are evaluated at the end of each block. The process of checking invariants is done via a special module called the [`InvariantsRegistry`](./invariants.md#invariant-registry). The value of the ivnariant should be equal to a predicted value defined in the module. Should the value be different than the predicted one, special logic defined in the invariant registry will be triggered (usually the chain is halted). This is useful to make sure no critical bug goes unoticed and produces long-lasting effects that would be hard to fix. 

Reviewer 5: ```suggestion
- With the module manager, register the [application's modules' invariants](./invariants.md). Invariants are variables (e.g. total supply of a token) that are evaluated at the end of each block. The process of checking invariants is done via a special module called the [`InvariantsRegistry`](./invariants.md#invariant-registry). The value of the invariant should be equal to a predicted value defined in the module. Should the value be different than the predicted one, special logic defined in the invariant registry will be triggered (usually the chain is halted). This is useful to make sure no critical bug goes unoticed and produces long-lasting effects that would be hard to fix. 
```

Reviewer 5: just a typo on `ivnariants`

- With the module manager, set the order of execution between the `InitGenesis`, `BegingBlocker` and `EndBlocker` functions of each of the [application's modules](#application-module-interface). Note that not all modules implement these functions.  

> - Set the application's [`InitChainer`](#initchainer) (used to initialize the application when it is first started), [`BeginBlocker`, `EndBlocker`](#beginblocker-and-endlbocker) (called at the beginning and the end of every block) and [`anteHandler`](#baseapp.md#antehandler) (used to handle fees and signature verification).  

Reviewer 5: It's a bit hard to read, maybe instead of parentheses around the description this can be transformed into a list

> - Set the application's [`InitChainer`](#initchainer) (used to initialize the application when it is first started), [`BeginBlocker`, `EndBlocker`](#beginblocker-and-endlbocker) (called at the beginning and the end of every block) and [`anteHandler`](#baseapp.md#antehandler) (used to handle fees and signature verification).  

Reviewer 4: ```suggestion
- Set the application's [`InitChainer`](#initchainer) (used to initialize the application when it is first started), [`BeginBlocker`, `EndBlocker`](#beginblocker-and-endblocker) (called at the beginning and the end of every block) and [`anteHandler`](#baseapp.md#antehandler) (used to handle fees and signature verification).  
```
